internal/tests/serialization/set.go

```diff
@@ -0,0 +1,212 @@
+				t.Run(stringValue(val), func(t *testing.T) {
+					s.runMarshalCorruptTest(t, marshal, val)
+				})
+				continue
+			}
+
+			if rt := reflect.TypeOf(val); rt.Kind() != reflect.Ptr {
+				unmarshalIn := newRef(val)
+				s.runUnmarshalCorruptTest(fmt.Sprintf("%T", val), t, unmarshal, unmarshalIn)
+			} else {
+				// Test unmarshal to (*type)(nil)
+				unmarshalIn := newRef(val)
+				s.runUnmarshalCorruptTest(fmt.Sprintf("%T**nil", val), t, unmarshal, unmarshalIn)
+
+				// Test unmarshal to &type{}
+				unmarshalInZero := newRefToZero(val)
+				s.runUnmarshalCorruptTest(fmt.Sprintf("%T**zero", val), t, unmarshal, unmarshalInZero)
+			}
+		}
+	})
+}
+
+func (s Set) runMarshalTest(t *testing.T, f func(interface{}) ([]byte, error), val interface{}) {
+	t.Run("marshal", func(t *testing.T) {
+
+		result, err := func() (d []byte, err error) {
+			defer func() {
+				if r := recover(); r != nil {
+					err = panicErr{err: r.(error), stack: debug.Stack()}
+				}
+			}()
+			return f(val)
+		}()
+
+		expected := bytes.Clone(s.Data)
+		if err != nil {
+			if !errors.As(err, &panicErr{}) {
+				err = errors.Join(marshalErr, err)
+			}
+		} else if !equalData(expected, result) {
+			err = unequalError{Expected: stringData(s.Data), Got: stringData(result)}
+		}
+
+		if isTypeOf(val, s.BrokenMarshalTypes) {
+			if err == nil {
+				t.Fatalf("expected to fail for (%T), but did not fail", val)
+			}
+			t.Skipf("skipped bacause there is unsolved problem")
+		}
+		if err != nil {
+			t.Error(err)
+		}
+	})
+}
+
+func (s Set) runUnmarshalTest(name string, t *testing.T, f func([]byte, interface{}) error, expected, result interface{}) {
+	t.Run(name, func(t *testing.T) {
+
+		expectedPtr := getPointers(result)
+
+		err := func() (err error) {
+			defer func() {
+				if r := recover(); r != nil {
+					err = panicErr{err: fmt.Errorf("%s", r), stack: debug.Stack()}
+				}
+			}()
+			return f(bytes.Clone(s.Data), result)
+		}()
+
+		if err != nil {
+			if !errors.As(err, &panicErr{}) {
+				err = errors.Join(unmarshalErr, err)
+			}
+		} else if !equalVals(expected, deReference(result)) {
+			err = unequalError{Expected: stringValue(expected), Got: stringValue(result)}
+		} else {
+			err = expectedPtr.Valid(result)
+		}
+
+		if isTypeOf(expected, s.BrokenUnmarshalTypes) {
+			if err == nil {
+				t.Fatalf("expected to fail for (%T), but did not fail", expected)
+			}
+			t.Skipf("skipped bacause there is unsolved problem")
+		}
+		if err != nil {
+			t.Error(err)
+		}
+	})
+}
+
+func (s Set) runMarshalCorruptTest(t *testing.T, f func(interface{}) ([]byte, error), val interface{}) {
+	_, err := func() (d []byte, err error) {
+		defer func() {
+			if r := recover(); r != nil {
+				err = panicErr{err: r.(error), stack: debug.Stack()}
+			}
+		}()
+		return f(val)
+	}()
+
+	if isTypeOf(val, s.BrokenMarshalTypes) {
+		if err != nil {
+			t.Fatalf("expected to not fail for (%T), but did fail", val)
+		}
+		t.Skipf("skipped bacause there is unsolved problem")
```

Comment:
Do I see it correctly that `BrokenMarshalTypes` for `RunCurruptTest` means that test does not fail for given type ?

Reply:
It seems that I got confused myself when I wrote this code and messages.
I've redone the code and messages for more clarity.